src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/resource/CreateInductionTest.kt

```diff
@@ -109,31 +109,33 @@ class CreateInductionTest : IntegrationTestBase() {
   }
 
   @Test
-  fun `should fail to create induction given database exception`() {
+  @Disabled("test is indeterminate due to the async nature of the coroutine - sometimes it will succeed meaning the controller has returned before the timeline database error; other times it will fail because the timeline database error happens before the controller has returned and the exception is propagated back")
+  fun `should create induction but fail to create timeline event given timeline database exception`() {
```

Comment:
The changes to this test are the exact same changes I made in this test in the ThreadPoolTaskExecutor based PR, and the changes were made for the same reason.

However, in this coroutine based implementation the test is indeterminate - sometimes it will pass which means the main request thread has completed and returned before the async timeline event database write has failed.
But other times the test will fail, which means the async timeline event database write failed before the main request thread completed, and the exception propagated back to the controller and its global exception handler.

Notwithstanding the fact that actually we should be validating our inputs better such that this scenario can never happen, the fact that it's indeterminate means that the timeline event service is not properly "fire and forget"

Also, its actually the source of a memory leak that I've read about online. If the main request thread has completed before the timeline event service fails and throws it's exception ..... where does that exception go? There is talk online about this as a memory leak, but I'm not sure of best approach and strategy for addressing it

Reply:
The problem is that by reusing the parent CoroutineScope you've bumped into [Structured Concurreny](https://kotlinlang.org/docs/coroutines-basics.html#structured-concurrency) - this is where Kotlin keeps track of all child coroutines and cancels both the parent and siblings if any coroutine fails.

Please see [this commit](https://github.com/ministryofjustice/hmpps-education-and-work-plan-api/pull/284/commits/8e3f23b2b9607f404e6a9cbafe6fe99497ecc0bf) - it creates a new CoroutineScope that is not linked to the parent scope and as such achieves the "fire and forget" behaviour you desire.

Reply:
Ah, I see. So we use a spring bean that is just for these "fire & forget" coroutine scopes! Got it! That makes sense 👍